File: JS/LeetC/linkedLists/oddEvenList.js
// Given the head of a singly linked list, group all the nodes with odd indices together followed by the nodes with even indices, and return the reordered list.

// The first node is considered odd, and the second node is even, and so on.

// Note that the relative order inside both the even and odd groups should remain as it was in the input.

function oddEvenList(head) {
	if (!head || !head.next) return head;

	let oddHead = head;
	let oddTail = head;

	let evenHead = head.next;
	let evenTail = head.next;

	while (evenTail && evenTail.next) {
		oddTail.next = evenTail.next;
		oddTail = oddTail.next;
		evenTail.next = oddTail.next;
		evenTail = evenTail.next;
	}

	oddTail.next = evenHead;
	return head;
}


let list1 = {
	val: 1,
	next: {
		val: 2,
		next: {
			val: 3,
			next: {
				val: 4,
				next: {
					val: 5,
					next: null,
				},
			},
		},
	},
};

console.log(oddEvenList(list1));

File: python/leetcode/linked-list/odd_even_list.py
# Given the head of a singly linked list, group all the nodes with odd indices together followed by the nodes with even indices, and return the reordered list.

# The first node is considered odd, and the second node is even, and so on.

from linked_list_leedcode import Node


class Solution:
    def oddEvenList(self, head):
        if head is None:
            return head

        odds = Node(0)
        evens = Node(0)
        oddsHead = odds
        evensHead = evens
        isOdd = True

        while head:
            if isOdd:
                odds.next = head
                odds = odds.next
            else:
                evens.next = evens
                evens = evens.next

            isOdd = not isOdd
            head = head.next

        evens.next = None
        odds.next = evensHead.next

        return oddsHead.next
